engine/collection/pusher/engine.go

```diff
@@ -53,81 +91,107 @@ func New(log zerolog.Logger, net network.EngineRegistry, state protocol.State, e
 // a potential processing error internally when done.
 func (e *Engine) Submit(channel channels.Channel, originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
+	engine.LogError(e.log, fmt.Errorf("pusher engine should only receive local messages on the same node"))
 }
 
 // ProcessLocal processes an event originating on the local node.
 func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(e.me.NodeID(), event)
-	})
+	ev, ok := event.(*messages.SubmitCollectionGuarantee)
+	if ok {
+		e.SubmitCollectionGuarantee(ev)
+		return nil
+	} else {
+		return fmt.Errorf("invalid message argument to pusher engine")
+	}
 }
 
 // Process processes the given event from the node with the given origin ID in
 // a blocking manner. It returns the potential processing error when done.
-func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, message any) error {
+	return fmt.Errorf("pusher engine should only receive local messages on the same node")
 }
 
-// process processes events for the pusher engine on the collection node.
-func (e *Engine) process(originID flow.Identifier, event interface{}) error {
-	switch ev := event.(type) {
-	case *messages.SubmitCollectionGuarantee:
-		e.engMetrics.MessageReceived(metrics.EngineCollectionProvider, metrics.MessageSubmitGuarantee)
-		defer e.engMetrics.MessageHandled(metrics.EngineCollectionProvider, metrics.MessageSubmitGuarantee)
-		return e.onSubmitCollectionGuarantee(originID, ev)
-	default:
-		return fmt.Errorf("invalid event type (%T)", event)
+// SubmitCollectionGuarantee adds a collection guarantee to the engine's queue
+// to later be published to consensus nodes.
+func (e *Engine) SubmitCollectionGuarantee(msg *messages.SubmitCollectionGuarantee) {
+	ok := e.queue.Push(msg)
+	if !ok {
+		engine.LogError(e.log, fmt.Errorf("failed to store collection guarantee in queue"))
+		return
 	}
+	e.notifier.Notify()
 }
 
-// onSubmitCollectionGuarantee handles submitting the given collection guarantee
-// to consensus nodes.
-func (e *Engine) onSubmitCollectionGuarantee(originID flow.Identifier, req *messages.SubmitCollectionGuarantee) error {
-	if originID != e.me.NodeID() {
-		return fmt.Errorf("invalid remote request to submit collection guarantee (from=%x)", originID)
-	}
-
-	return e.SubmitCollectionGuarantee(&req.Guarantee)
-}
-
-// SubmitCollectionGuarantee submits the collection guarantee to all consensus nodes.
-func (e *Engine) SubmitCollectionGuarantee(guarantee *flow.CollectionGuarantee) error {
+// publishCollectionGuarantee publishes the collection guarantee to all consensus nodes.
```

Comment:
Could you please add error return documentation to this function. If it might return an expected sentinel error (including from sub-calls into `Identities()` and `Publish()`), we should list those possible error returns. Otherwise, we should add the line `// No errors expected during normal operation.`

If you haven't already, take a look at the godoc references for [`Identities()`](https://github.com/onflow/flow-go/blob/448fa4b1c777e0a0e6b7c7c591df476e8ac6f31b/state/protocol/snapshot.go#L48-L63) and [`Publish()`](https://github.com/onflow/flow-go/blob/1ba41690954621d15d59dcabcc8f0dfc0b56f1bd/network/conduit.go#L31-L39) and think about what errors they might return, and which of those are expected vs unexpected in the context of executing `publishCollectionGuarantee`. 

In particular, `Identities()` does have a listed error return value -- `state.ErrUnknownSnapshotReference` -- but in the context here, it is _not_ expected. Why is that?

Reply:
For `Identities()`, I believe since we're using the `Final()` (latest finalized in cache) snapshot, we expect to already have a valid snapshot reference / queryable block in the state, so expect `Identities()` on it to never error.

From what I can tell `Publish()` only returns an error if the conduit channel has been closed (in which case I would assume we can't continue), or other ["benign" errors](https://github.com/onflow/flow-go/blob/1ba41690954621d15d59dcabcc8f0dfc0b56f1bd/network/p2p/node/libp2pNode.go#L429-L430) when the constructed message is invalid in some way (pretty sure this is not expected) or some failure in the underlying libp2p (I assume also not expected).